vowpalwabbit/cats.cc

```diff
@@ -0,0 +1,179 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+
+#include "cats.h"
+#include "parse_args.h"
+#include "err_constants.h"
+#include "api_status.h"
+#include "debug_log.h"
+
+// Aliases
+using VW::LEARNER::single_learner;
+using std::endl;
+using VW::cb_continuous::continuous_label;
+using VW::cb_continuous::continuous_label_elm;
+using VW::config::make_option;
+using VW::config::option_group_definition;
+using VW::config::options_i;
+
+// Enable/Disable indented debug statements
+VW_DEBUG_ENABLE(false)
+
+// Forward declarations
+namespace VW
+{
+  void finish_example(vw& all, example& ec);
+}
+
+namespace VW { namespace continuous_action { namespace cats {
+  ////////////////////////////////////////////////////
+  // BEGIN cats reduction and reduction methods
+  struct cats
+  {
+    cats(single_learner* p_base);
+
+    int learn(example& ec, api_status* status);
+    int predict(example& ec, api_status* status);
+
+  private:
+    single_learner* _base = nullptr;
+  };
+
+  // Pass through
+  int cats::predict(example& ec, api_status*)
+  {
+    VW_DBG(ec) << "cats::predict(), " << features_to_string(ec) << endl;
+    _base->predict(ec);
+    return error_code::success;
+  }
+
+  // Pass through
+  int cats::learn(example& ec, api_status* status = nullptr)
+  {
+    assert(!ec.test_only);
+    predict(ec, status);
+    VW_DBG(ec) << "cats::learn(), " << to_string(ec.l.cb_cont) << features_to_string(ec) << endl;
+    _base->learn(ec);
+    return error_code::success;
+  }
+
+  cats::cats(single_learner* p_base) : _base(p_base) {}
+
+  // Free function to tie function pointers to reduction class methods
+  template <bool is_learn>
+  void predict_or_learn(cats& reduction, single_learner&, example& ec)
+  {
+    api_status status;
+    if (is_learn)
+      reduction.learn(ec, &status);
+    else
+      reduction.predict(ec, &status);
+
+    if (status.get_error_code() != error_code::success)
```

Comment:
status.isError() or  ! status.isSuccess()

or something like that?

Reply:
Let's disuss in error code PR.  This will be replaced by a macro.